### ISSUE TITLE
Niivue + ONNX clientside

### ISSUE THREAD

**Issue by iishiishii** [niivue/niivue#663]
## Goal:

The clientside enables users to perform image preprocessing using niimath (output as NVImage for other computation) and execute a segmentation model (output drawings). Users can refine and improve ROI labels for model training.

## Features

**Navbar**

- File : Upload images, Download drawing/image in the current canvas
- View: Clip plane, Radiological, Crosshair, Show colorbar, Colormap
- Opacity and Intensity range stay more explicit for their frequent use.

**Canvas**:

I find multiplanar view more useful but having them all in one canvas seems to be difficult to get the slice coordinates.

**Computing**

- Preprocessing: niimath takes DT_UNSIGNED_CHAR for the datatype of input array (I checked an example in niimath repo, it says to support dataypes 2,4,16. I didn't get 4 and 16 to work but I dont think it's an issue https://github.com/rordenlab/niimath/blob/f644a9738ed1f014acd39818dae153f7a6f225e3/src/niimath.js#L91)
- Pretrained Model (WIP): SAM encoder takes Float32 dims array, the decoder takes Float32 + click coord + few more inputs and output Float32 array.
If possible, when user selects axial view, it gets all the slices in that view and passes to the encoder (might be very laggy as the encoder is 10 times larger than the decoder). It seems possible to have efficient multiple computations using web worker. To get the model output in the canvas as drawing for refining, when clicking on the canvas, the decoder then get the mask for all the slices (so we have 3d output). 
I have some issues that need your advice @neurolabusc or others in the team:
1. I think @shrutivarade has a smart solution to get 2d image data from canvas context so the coordinates from mousedown event makes sense. But can we somehow use the mouse coordinate to get the image slice? I think there might be something like that for draw function.
2. The model needs row major order array but it seems that NVImage buffer is column major array? 

## Potential UI
![niivue](https://github.com/niivue/niivue/assets/59903692/7416289d-aa33-4c82-9bc2-88077c19c47b)

@stebo85 @haehn

**Comment by hanayik** [comment]
@iishiishii , Thanks for opening this issue. I think I understand what you're after. I don't think it makes sense to bundle all of this into NiiVue (I'm not sure if you were suggesting that, but just thought I'd share the thought). 

NiiVue is a library for parsing and viewing medical imaging data in modern browsers using webgl2.0. We have a focus on brain imaging since that's the field many of the developers work in. 

As you point out, NiiVue can be extended to work with our WASM niimath module for preprocessing, but that module is not bundled in by default. The two modules are peers, but one does not depend on the other. 

I think we should use that same approach when thinking of interacting with ONNX models. We can **probably add features** to NiiVue that make it easier to interact with user specified ONNX models, but the "glue" that connects NiiVue to the model could be handled by the developers that import NiiVue into their projects. 

I think a first pass at getting NiiVue to interact with an ONNX model like the one you describe would be to add some helper functions that allow sending and receiving slice data to/from the model code. 

Perhaps when interacting with segmentation models like this, developers can activate an empty drawing by default, and then as the model outputs new data, the relevant slices in the drawing can be updated to display results overlaid on the background image. 

Does your proposed model accept 3D inputs, or only 2D?

**Comment by neurolabusc** [comment]
@iishiishii  these are great points, and I think your mock up looks great. I would suggest:
 1. We add core functions and call backs that make reading and writing data to NiiVue simpler and more interactive with mouse gestures. I can take the lead on this.
 2. I would suggest you develop a modular component for ONNX that wraps NiiVue, in the same way that [we have an example of niimath WASM linking with NiiVue](https://github.com/niivue/niivue-niimath). I have invited you to be a member of the NiiVue team, and would invite you to make a new project in the NiiVue tree (though you can keep it in your own organization if you prefer).
 3. I think it would be great if we could work with @haehn and @FarfallaHu, @Mmasoud1  and @sergeyplis of the [brainchop](https://github.com/neuroneural/brainchop) with their impressive [Papaya-based livedemo](http://brainchop.org/). It would be nice if we could come up with a clean API for a lot of these interactions.
 4. I do hope we can provide the correct hooks and interactions to do everything as one WebGL2 canvas. This will require less resources and make it easier to auto-adapt to different screen sizes like computers and phones.

**Comment by haehn** [comment]
Thanks for starting this discussion! We suggest enabling the following 3 core functionalities in niivue:

1. Data Read Access
- to retrieve image (+segmentation) data as chunks similar to numpy slicing
- if possible, different data types and array order should be supported but i think it is fine if not since this could be done outside of niivue

2. Interaction Access
- especially placement of seeds/selection of ROI by easily enabling the built-in widgets

3. Data Write Access
- to add image and segmentation layers back to the niivue scene with proper re-drawing etc.
- this could later be expanded to fibers, mesh overlays etc.

Such a generic API would allow processing independently of ONNX.js or any other framework (Tensorflow.js...) and enable flexibility for developing external plugins.

**Comment by neurolabusc** [comment]
@haehn thanks for creating a concrete set of features and milestones. I am happy to work on the NiiVue side of this, and create minimal live demo examples that your team and the neurodesk team can exploit.

**Comment by gaiborjosue** [comment]
Hello! Regarding the previous discussion about the API and new requested features, we have a JSdoc as an overview of the functionality:

```
/**
 * @param x1 {int, null} the bbox start x coordinate, if null, set to 0
 * @param x2 {int, null} the bbox end x coordinate, if null, set to volume dimension
 * @param y1 {int, null} the bbox start y coordinate, if null, set to 0
 * @param y2 {int, null} the bbox end y coordinate, if null, set to volume dimension
 * @param z1 {int, null} the bbox start z coordinate, if null, set to 0
 * @param z2 {int, null} the bbox end z coordinate, if null, set to volume dimension
 * @param normalized {bool} to return original voxel range or 0..1
 * @param datatype {Uint8Array, Float16Array…, null} the encoding of the voxels, if null, the original encoding
 * @param orientation {String, null} if ‘RAS’ or ‘LPS’ honor this or return niivue ‘RAS’ standard if null
 * @returns {Array} The subvolume data within a range of specified coordinates.
 */
nv.get_volume_data ( x1, x2, y1, y2, z1, z2, normalized, datatype, orientation )

 /**
 * @param x1 {int, null} the bbox start x coordinate, if null, set to 0
 * @param x2 {int, null} the bbox end x coordinate, if null, set to volume dimension
 * @param y1 {int, null} the bbox start y coordinate, if null, set to 0
 * @param y2 {int, null} the bbox end y coordinate, if null, set to volume dimension
 * @param z1 {int, null} the bbox start z coordinate, if null, set to 0
 * @param z2 {int, null} the bbox end z coordinate, if null, set to volume dimension
 * @returns {Array} The label data for the specified subvolume as an array.
 */
nv.get_label_data(x1, x2, y1, y2, z1, z2)
```

These functions will allow for retrieving both volume data and label data. Thanks for being open to integrating the new functionalities.

**Comment by neurolabusc** [comment]
1. It seems like normalized as a `bool` is too limited. Consider a 16-bit image based on a 12-bit ADC where voxels have intensity range of 0..4095, but the user has adjusted the cal_min and cal_max to be 24..1024 (e.g. window center = 524, window width = 1000). It seems like the options are:
 - raw (0..4095)
 - rawNormalized (0..4095 -> 0..1)
 - rawWindowed (24..1024 -> 0..1)

2. When the user chooses normalized, the output is the full dynamic range of the datatype, or the full positive dynamic range. Consider the user wants the values normalized but with `null` (native) datatype where the data was INT16 (full dynamic range -32768..32767) - is the expectation that the result scales 0..1 to 0..32767 (positive 15 bits) or the full 16-bit -32768..32767?

**Comment by gaiborjosue** [comment]
We would suggest using the full range but are happy to listen to the community here.

**Comment by hanayik** [comment]
@gaiborjosue, do you expect that niivue will be responsible for any resizing prior to inputting the extracted subvolumes into a model? Perhaps we should leave that up to model creators, or have it as a processing step outside of niivue. 

For example, if a subvolume has dimensions `512, 512, 1` (a single slice from a CT head image loaded in NiiVue), but the model expects 2D slices with dimensions `128, 128, 1`, where will the resize operation happen? Model? or as an extra step? does that make sense? or is it a non-problem?

**Comment by gaiborjosue** [comment]
Hello @hanayik, in our opinion, it should be up to the model creators. Thanks!

**Comment by neurolabusc** [comment]
The [NiiVue live demo page](https://niivue.github.io/niivue/) now includes a "Machine Learning and AI helper functions" section with three live demos:
 - [label voxel clusters](https://niivue.github.io/niivue/features/label_clusters.html) identifies [connected components](https://github.com/seung-lab/connected-components-3d) which is useful for cleaning up machine learning labelling. It uses the [createConnectedLabelImage()](https://github.com/niivue/niivue/blob/853f280c4989fec11f5936de60d2475d7a270160/src/niivue/index.ts#L6060) function.
 - The [conform](https://niivue.github.io/niivue/features/conform.html) live demo ports the FreeSurfer/FastSurfer conform function to pure JavaScript. It uses the new [conform()](https://niivue.github.io/niivue/devdocs/classes/Niivue.html#conform) function.
 - The [slab selection](https://niivue.github.io/niivue/features/slab_selection.html) demo is most relevant to this issue. This demonstrates howw to clone an existing volume or create a new volume from scratch. It shows how one can read and write either all or parts (slabs) of a voxel based volume. The key new functions are [getVolumeData](
https://niivue.github.io/niivue/devdocs/classes/NVImage.html#getVolumeData) and [setVolumeData](https://niivue.github.io/niivue/devdocs/classes/NVImage.html#setVolumeData).

The getVolumeData() and setVolumeData() provide an easy way to read and write slabs of voxels. The X,Y,Z coordinates for selecting the row, columns and slices are always specificied in [RAS](https://www.nitrc.org/plugins/mwiki/index.php/dcm2nii:MainPage#Spatial_Coordinates) coordinates, regardless of the order of data as stored on disk.

Happy to get any feedback on these new experimental features. Since we are open to feedback, do be aware that these new function calls may be changed in upcoming stable releases.
